draft-irtf-cfrg-rsa-blind-signatures.md

```diff
@@ -338,6 +352,119 @@ Steps:
    raise "invalid signature" and stop
 ~~~
 
+## External Application Interface {#salted-interface}
+
+This section presents an application interface for blinding, finalizing, and verifying
+messages that is build on the internal functions described in {{generation}}. This
+interface injects additional entropy into application messages by choosing a random
+salt of length 32 bytes, prepending the salt to the input message, and then invoking
+the internal functions in {{generation}}. Note that this only changes what is passed
+to rsabssa_blind and rsabssa_finalize, as the application message is not provided as
+input to rsabssa_blindsign.
+
+Applciations that provide high-entropy input messages can expose the internal
+rsabssa_blind and rsabssa_finalize directly, as the additional message randomization
+does not offer security advantages. See {{apis}} and {{message-entropy}} for more information.
+
+### Salted Blind
+
+rsabssa_salted_blind invokes rsabssa_blind with a salted input message and outputs the
+blinded message to be sent to the server and the corresponding inverse, both encoded
+as octet strings, as well as the fresh message salt, which is 32 random bytes.
+
+~~~
+rsabssa_salted_blind(pkS, msg)
+
+Parameters:
+- kLen, the length in octets of the RSA modulus n
+- kBits, the length in bits of the RSA modulus n
+- HF, the hash function used to hash the message
+- MGF, the mask generation function
+
+Inputs:
+- pkS, server public key (n, e)
+- msg, message to be signed, an octet string
+
+Outputs:
+- blinded_msg, an octet string of length kLen
+- inv, an octet string of length kLen
+- msg_salt, an octet string of length 32 bytes
+
+Errors:
+- "message too long": Raised when the input message is too long.
+- "encoding error": Raised when the input message fails encoding.
+- "invalid blind": Raised when the inverse of r cannot be found.
+
+Steps:
+1. msg_salt = random(32)
+2. salted_msg = msg_salt || msg
+3. blinded_msg, inv = blind(pkS, msg)
+4. output msg_salt, blinded_msg, inv
+~~~
+
+### Salted Finalize
+
+rsabssa_salted_finalize invokes rsabssa_finalize directly with the salted
+message and outputs the result.
+
+~~~
+rsabssa_salted_finalize(pkS, msg, blind_sig, inv)
+
+Parameters:
+- kLen, the length in octets of the RSA modulus n
+
+Inputs:
+- pkS, server public key (n, e)
+- msg, message to be signed, an octet string
+- msg_salt, the 32 octets random salt used to salt the message
+- blind_sig, signed and blinded element, an octet string of
+  length kLen
+- inv, inverse of the blind, an octet string of length kLen
+
+Outputs:
+- sig, an octet string of length kLen
+
+Errors:
+- "invalid signature": Raised when the signature is invalid
+- "unexpected input size": Raised when a byte string input doesn't
+  have the expected length.
+
+Steps:
+1. salted_msg = msg_salt || msg
```

Comment:
salted_msg = HF(salt_msg || msg ) as defined lower

Reply:
Sticking with what we have since the hash function here is redundant.